students/Kate_Koehn/Session06/mailroom_4.py

```diff
@@ -0,0 +1,172 @@
+    report.append("{:-<60}".format(""))
+    for donor_name, gifts in donors.items():
+        total = sum(gifts)
+        num_gifts = len(gifts)
+        avg_gift = total / num_gifts
+        report.append("{:<22} ${:<13.2f} {:<10} ${:.2f}".format(donor_name, total, num_gifts, avg_gift))
+    return "\n".join(report)
+
+
+def email(donor_name, donation_amount):
+    """creates individual thank-you email for a new donation"""
+    message = ("\n\nDear {},"
+               "\n\nThanks for your money! Your donation of ${:.2f} will be summarily used to buy me beer.\n\n"
+               "Kind regards,\nKate Koehn").format(donor_name, donation_amount)
+    file_name = "{}.txt".format(donor_name)
+    with open(file_name, "w") as f:
+        f.write(message)
+
+
+def generate_letter(donor_name, donation_amount):
+    """generates a thank you letter for the donation made"""
+    return dedent('''Dear {!s:s},
+
+    Thank you for your crappy donation of ${:.2f}. It will be used to buy the cheapest beer possible, because you couldn't donate enough for us to buy anything decent. We'll be lucky if we get Natty Ice.
+
+    Cheers,
+        Kate Koehn
+    '''.format(donor_name, donation_amount))
+
+
+def get_donor_key(donor_name):
+    """looks up key in donor dict"""
+    donor_key = donor_name.strip()
+    for key in donors.keys():
+        if donor_key == key:
+            return True
+    return False
+
+
+def add_new_donor(donor_name):
+    name = donor_name.strip()
+    if name not in donors.keys():
+        donors[name] = []
+    return donor_name
+
+
+def add_donation_to_donor(donor_name, donation):
+    donors[donor_name].append(donation)
+
+
+def check_donation(donation):
+    donation = float(donation)
+    if math.isnan(donation) or math.isinf(donation) or round(donation, 2) <= 0.00:
+        raise ValueError
+    return donation
+
+
+def record_new_donation(donor_name):
+    """prompts user for a new donation amount and store it in the donor dict"""
+    while True:
+        donation = input("Please enter the donation amount or type 'menu' to return to the main menu: ").strip()
+
+        if donation == "menu":
+            return
+        try:
+            donation = check_donation(donation)
+        except ValueError:
+            print("You have tried to donate an invalid amount. Please try again.\n")
+        else:
+            break
+
+    # check if donor already exists in donors dict, if not, append name and donation
+    add_new_donor(donor_name)
+    add_donation_to_donor(donor_name, donation)
+    email(donor_name, donation)
+
+
+def thanks():
+    """sub-menu to record a new donation and thank the user for their donation"""
+    while True:
+        print("\n\n~ Thanking Your Donors ~\n")
+        donor_name = input("Enter a donor's name or 'list' to see all donors or 'menu' to exit to main menu: ").strip()
+
+        if donor_name == "list":
+            for name in donors:
```

Comment:
seems trivial, but s "list_donors" function may be called for -- particular if you might want to make a fancier listing at some point.